libs/SmartSync/src/com/salesforce/androidsdk/smartsync/util/SoqlSyncDownTarget.java

```diff
@@ -130,6 +138,80 @@ public JSONArray continueFetch(SyncManager syncManager) throws IOException, JSON
         return records;
     }
 
+    @Override
+    public Set<String> getListOfRemoteIds(SyncManager syncManager, Set<String> localIds) {
+        if (localIds == null) {
+            return null;
+        }
+        final String idFieldName = getIdFieldName();
+        final Set<String> remoteIds = new HashSet<String>();
+
+        /*
+         * Compute how many SOQL queries are required to get the IDs.
+         * SOQL has a limit of 2000 records per query, so we need to break it up.
+         */
+        final List<Set<String>> uberQueryList = new ArrayList<Set<String>>();
```

Comment:
```SOQL``` has a limit of ```2000``` records, so we need to break the set into subsets and send them as separate ```SOQL``` queries.